enhancements/node-pull-credentials/pull-credentials.md

```diff
@@ -0,0 +1,253 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Allow OpenShift users to import and use images from any registry configured
+during or after the cluster installation by sharing node's pull credentials
+with `openshift-api`, `builder` and `image-registry` pods.
+
+## Motivation
+
+To have image stream, image registry and builds working closely to node avoids
+complexity and redundancy as each component won't need to mount different
+Secrets to have access to information that is already present on the node's
+filesystem.
+
+OpenShift should seamlessly use the cluster-wide pull secret provided during
+installation in builds, imagestream imports and pull-through operations. This
+is particularly important for images pulled from `registry.redhat.io`, which
+requires a pull secret.
+
+Today pull secrets provided during cluster install are available on the
+node's filesystem. If user attempts to import an image stream or pull-through
+from these locations, OpenShift fails as none of `openshift-api`, `builder` or
+`image-registry` use the credentials provided during the installation.
+
+### Goals
+
+- Allow users to use container images from any registry provided during or
+  after cluster installation.
+- Allow users to run `builds` based on images hosted in any registry provided
+  during or after cluster installation without providing any extra credentials.
+- Allow `image-registry` to execute pull-through using the node's pull
+  credentials.
+
+### Non-Goals
+
+- Change the way users manage cluster pull credentials.
+  * Additional credentials may still be provided through secrets on namespace.
+- Use Node mirroring config.
+  * To use config files files such as `/etc/containers/registries.conf` is not
+  part of this proposal.
+- Use Node CAs.
+  * This should be done in the future through another enhancement proposal as
+  we currently have different sources of truth when it comes to CAs (different
+  methods of providing them to different parts of the codebase).
+
+## Proposal
+
+### User Stories
+
+As an OpenShift user
+I want to be able to import imagestreams from any image registry that the
+cluster nodes can pull images from
+so that I can use them without manually creating any extra credentials.
+
+As an OpenShift user
+I want to be able to use, during builds, images from any image registry that
+the cluster nodes can pull images from
+so that I can use these images as base for my own images.
+
+As an OpenShift user
+I want to be able to pull-through images from any registry that the cluster
+nodes can pull images from
+so that I can use image registry cache to speed-up my builds.
+
+### Implementation Details/Notes/Constraints
+
+This is the node's filesystem path we are taking into account:
+
+- `/var/lib/kubelet/config.json` contains the node's pull secret.
```

Comment:
can this file change? E.g. a user can ssh into the machine and append secrets there.

Reply:
This has influence on https://github.com/openshift/cluster-openshift-apiserver-operator/pull/284, i.e. we need to reread that file. Is that the case?

Reply:
it can be changed via the MCO which causes node reboots so rereading is moot.

if the MCO is ever updated to allow node filesystem changes w/o rebooting the nodes, we'd have to setup a watch on the mounted file in our pod.

I believe these steps are what lead to the /var/lib/kubelet/config.json being updated:
https://docs.openshift.com/container-platform/4.3/openshift_images/managing-images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets

Reply:
Pull secrets are read in a "per image stream import request", we don't keep it on memory.